swarm/storage/hasherstore.go

```diff
@@ -107,16 +110,40 @@ func (h *hasherStore) Get(ctx context.Context, ref Reference) (ChunkData, error)
 // Close indicates that no more chunks will be put with the hasherStore, so the Wait
 // function can return when all the previously put chunks has been stored.
 func (h *hasherStore) Close() {
-	close(h.closed)
+	close(h.doneC)
 }
 
 // Wait returns when
 //    1) the Close() function has been called and
 //    2) all the chunks which has been Put has been stored
 func (h *hasherStore) Wait(ctx context.Context) error {
-	<-h.closed
-	h.wg.Wait()
-	return nil
+	defer close(h.quitC)
+	var nrStoredChunks uint64 // number of stored chunks
+	var done bool
+	doneC := h.doneC
+	for {
+		select {
+		// if context is done earlier, just return with the error
+		case <-ctx.Done():
+			return ctx.Err()
+		// doneC is closed if all chunks have been submitted, from then we just wait until all of them are also stored
+		case <-doneC:
+			done = true
+			doneC = nil
```

Comment:
What's the reason to set `doneC = nil` and not close it?

Reply:
Nil channel is blocking forever. When doneC channel is closed, this select will step into this case, and being in the for loop it will create a very tight loop by stepping into it in every for loop iteration. Setting doneC to nil will prevent this, as nil channels are blocked, but other cases are used for control flow, in this case h.errC and ctx.Done(). The if statement in this for loop is taking care that we exit the loop and function in the happy path, and the if in h.errC case will exit the loop on error.